examples/flushing/flushing_test.go

```diff
@@ -0,0 +1,39 @@
+package main
+
+import (
+	"flag"
+	"testing"
+
+	"go.uber.org/goleak"
+	"k8s.io/klog/v2"
+)
+
+func main() {
+	klog.InitFlags(nil)
+
+	// By default klog writes to stderr. Setting logtostderr to false makes klog
+	// write to a log file.
+	flag.Set("logtostderr", "false")
+	flag.Set("log_file", "myfile.log")
+	flag.Parse()
+
+	// Info writes the first log message. When the first log file is created,
+	// a flushDaemon is started to frequently flush bytes to the file.
+	klog.Info("nice to meet you")
+
+	// klog won't ever stop this flushDaemon. To exit without leaking a goroutine,
+	// the daemon can be stopped manually.
+	klog.StopFlushDaemon()
+
+	// After you stopped the flushDaemon, you can still manually flush.
+	klog.Info("bye")
+	klog.Flush()
+}
+
+func TestLeakingFlushDaemon(t *testing.T) {
+	// goleak detects leaking goroutines.
+	defer goleak.VerifyNone(t)
```

Comment:
Not a problem with the code, I am just wondering: is this check racy?

The usual pattern for shutting down a goroutine is what you have implemented: the goroutine itself notifies the main goroutine that it is done, then returns from its function. But if the main goroutine then immediately wakes up (can that happen?) and checks for the existence of the goroutine, it might still find it because the goroutine is still in the process of terminating itself (or getting terminated).

In other languages there is an explicit "join thread", but not in Go. That's not normally a problem because once the goroutine function is ready to return, it doesn't do anything anymore and whether it still exists doesn't matter. But here it does :man_shrugging:

Reply:
Good question, I've never thought about this. It seems like they just sleep and retry if there are still goroutines running. https://github.com/uber-go/goleak/blob/421b885faf413a221d1a86a17684fb3c45bd35c5/leaks.go#L60-L67

Reply:
So it looks like this can happen, and goleak deals with it. Good :smile: